services/QuillLMS/app/controllers/snapshots_controller.rb

```diff
@@ -0,0 +1,168 @@
+      current_start: proc { DateTime.current.beginning_of_month - 1.month },
+      current_end: proc { DateTime.current.beginning_of_month },
+    }, {
+      value: 'this-year',
+      name: 'This year',
+      previous_start: proc { DateTime.current.beginning_of_year - 1.year },
+      current_start: proc { DateTime.current.beginning_of_year },
+      current_end: proc { DateTime.current },
+    }, {
+      value: 'last-year',
+      name: 'Last year',
+      previous_start: proc { DateTime.current.beginning_of_year - 2.years },
+      current_start: proc { DateTime.current.beginning_of_year - 1.year },
+      current_end: proc { DateTime.current.beginning_of_year },
+    }, {
+      value: 'all-time',
+      name: 'All time',
+      previous_start: proc { DateTime.new(2010,1,1) }, # This is well before anydata exists in our system, so works for "all time"
+      current_start: proc { DateTime.new(2010,1,1) }, # This is well before anydata exists in our system, so works for "all time"
+      current_end: proc { DateTime.current },
+    }, {
+      value: Snapshots::CacheKeys::CUSTOM_TIMEFRAME_NAME,
+      name: 'Custom',
+      previous_start: :custom_previous_start,
+      current_start: :custom_current_start,
+      current_end: :custom_current_end
+    }
+  ]
+
+  GRADE_OPTIONS = [
+    {value: "Kindergarten", name: "Kindergarten"},
+    {value: "1", name: "1st"},
+    {value: "2", name: "2nd"},
+    {value: "3", name: "3rd"},
+    {value: "4", name: "4th"},
+    {value: "5", name: "5th"},
+    {value: "6", name: "6th"},
+    {value: "7", name: "7th"},
+    {value: "8", name: "8th"},
+    {value: "9", name: "9th"},
+    {value: "10", name: "10th"},
+    {value: "11", name: "11th"},
+    {value: "12", name: "12th"},
+    {value: "University", name: "University"},
+    {value: "Other", name: "Other"}
+  ]
+
+  def count
+    response = Rails.cache.read(cache_key)
+
+    return render json: response if response
+
+    previous_timeframe_start, current_timeframe_start, timeframe_end = calculate_timeframes
+
+    Snapshots::CacheSnapshotCountWorker.perform_async(snapshot_params[:query],
+      current_user.id,
+      {
+        name: snapshot_params[:timeframe],
+        previous_start: previous_timeframe_start,
+        current_start: current_timeframe_start,
+        current_end: timeframe_end
+      },
+      snapshot_params[:school_ids],
+      snapshot_params[:grades])
+
+    render json: { message: 'Generating snapshot' }
+  end
+
+  def options
+    render json: {
+      timeframes: TIMEFRAMES.map { |timeframe| format_timeframe_option(timeframe) },
+      schools: Snapshots::SchoolsOptionsQuery.run(current_user.id),
+      grades: GRADE_OPTIONS
+    }
+  end
+
+  private def cache_key
+    Snapshots::CacheKeys.generate_key(snapshot_params[:query],
+      current_user.id,
+      {
+        name: snapshot_params[:timeframe_name],
+        custom_start: snapshot_params[:timeframe_custom_start],
```

Comment:
Are we sure these custom timeframes will be `nil` when passed in from the frontend? E.g. if a user starts playing with the calendar select, then selects a set timeframe.